controllers/capabilities/tenant_controller.go

```diff
@@ -95,119 +107,116 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 		reqLogger.V(1).Info(string(jsonData))
 	}
 
-	masterAccessToken, err := r.fetchMasterCredentials(tenantR)
-	if err != nil {
-		reqLogger.Error(err, "Error fetching master credentials secret")
-		// Error reading the object - requeue the request.
-		return ctrl.Result{}, err
+	// reconcileErr can be an error of different types - generic k8s err, isInvalid, waitError, isOrphaned
+	// different errors are handled differently when it comes to requeueing the reconciliation loop
+	statusReconciler, reconcileErr := r.specReconciler(tenantR, reqLogger)
+
+	// statusUpdateError can be an error when making the update call to the resource
+	res, statusUpdateErr := statusReconciler.Reconcile()
+	if statusUpdateErr != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update tenantCR: %w", statusUpdateErr)
 	}
 
-	insecureSkipVerify := controllerhelper.GetInsecureSkipVerifyAnnotation(tenantR.GetAnnotations())
-	portaClient, err := controllerhelper.PortaClientFromURLString(tenantR.Spec.SystemMasterUrl, masterAccessToken, insecureSkipVerify)
-	if err != nil {
-		reqLogger.Error(err, "Error creating porta client object")
-		// Error reading the object - requeue the request.
-		return ctrl.Result{}, err
+	// Requeue if either object or status were updated
+	if res.Requeue {
+		return reconcile.Result{}, nil
 	}
 
-	// Tenant has been marked for deletion
-	if tenantR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
-		existingTenant, err := controllerhelper.FetchTenant(tenantR.Status.TenantId, portaClient)
-		if err != nil {
-			return ctrl.Result{}, err
+	if reconcileErr != nil {
+		// On validation error, do not retry since there's something wrong with the spec
+		if helper.IsInvalidSpecError(reconcileErr) {
+			reqLogger.Info("ERROR", "spec validation error", reconcileErr)
+			r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "Invalid Spec", "%v", reconcileErr)
+			return ctrl.Result{}, nil
 		}
 
-		// delete tenantCR if tenant is present in 3scale
-		if existingTenant != nil {
-			// do not attempt to delete tenant that is already scheduled for deletion
-			if existingTenant.Signup.Account.State != scheduledForDeletionState {
-				err := portaClient.DeleteTenant(tenantR.Status.TenantId)
-				if err != nil {
-					r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "Failed to delete tenant", "%v", err)
-					return ctrl.Result{}, err
-				}
-			} else {
-				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion", "tenantID", existingTenant.Signup.Account.ID)
-			}
+		// On wait error, retry
+		if helper.IsWaitError(reconcileErr) {
+			reqLogger.Info("ERROR", "wait error", reconcileErr)
+			return ctrl.Result{Requeue: true}, nil
 		}
 
-		controllerutil.RemoveFinalizer(tenantR, tenantFinalizer)
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
+		// On orphan error, do not retry
+		if helper.IsOrphanSpecError(reconcileErr) {
```

Comment:
orphan state is conceptually between custom resources. Some custom resources may be children of other custom resources (like Account and application). 

For the tenant,  I do not thing this applies.